File: src/main/java/com/manifestors/shinrai/client/setting/settings/IntegerSetting.java
package com.manifestors.shinrai.client.setting.settings;

import com.manifestors.shinrai.client.setting.Setting;
import lombok.Getter;

public class IntegerSetting extends Setting<Integer> {

    @Getter
    private int current;
    @Getter
    private final int max;
    @Getter
    private final int min;
    private final int step;

    public IntegerSetting(String name, int current, int max, int min, int step) {
        super(name, current);
        this.current = current;
        this.max = max;
        this.min = min;
        this.step = step;
    }

    public void incrementCurrent() {
        current += step;

        if (current >= max) {
            current = max;
        }
    }

    public void decrementCurrent() {
        current -= step;

        if (current <= min) {
            current = min;
        }
    }

}

File: src/main/kotlin/com/manifestors/shinrai/client/setting/settings/IntegerSetting.kt
/*
 *
 * Copyright © 2026 Manifestors Open Source
 * License: GPL-3.0
 * All code in this project is the property of the Manifestors Open Source team
 * and its contributors. If you use this code in any project, please provide proper attribution
 * and release your project under the GPL-3.0 license as well.
 * For more details, see: https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 */

package com.manifestors.shinrai.client.setting.settings

import com.manifestors.shinrai.client.setting.Setting

class IntegerSetting(
    name: String,
    current: Int,
    val min: Int,
    val max: Int,
    val step: Int
) : Setting<Int>(name, current) {

    var currentValue: Int = current
        set(value) {
            field = value.coerceIn(min, max)
            super.current = field
        }

    fun change(by: Int) {
        currentValue += by
    }

    fun increment() = change(step)
    fun decrement() = change(-step)

}
